### PR TITLE
CB-3563 cloudbeaver preference store

### PR DESCRIPTION
**Pull request by alexander-skoblikov**

### CODE REVIEW

File: webapp/packages/core-settings/src/SettingsScopeType.ts [hunk 1]
```diff
@@ -6,4 +6,4 @@
  * you may not use this file except in compliance with the License.
  */
 
-export type SettingsScopeType = 'plugin' | 'core';
+export type SettingsScopeType = 'plugin' | 'core' | 'sql';
```
Comment (Wroud) [inline webapp/packages/core-settings/src/SettingsScopeType.ts]: maybe `'plugin' | 'core' | string` will be better

File: webapp/packages/plugin-codemirror6/src/ReactCodemirror.tsx [hunk 2]
```diff
@@ -155,12 +155,13 @@ export const ReactCodemirror = observer<IReactCodeMirrorProps, IEditorRef>(
     useLayoutEffect(() => {
       if (view) {
         const transaction: TransactionSpec = { annotations: [External.of(true)] };
+        const isCursorInDoc = cursor && cursor.anchor > 0 && cursor.anchor < view.state.doc.length;
 
         if (value !== undefined && value !== view.state.doc.toString()) {
           transaction.changes = { from: 0, to: view.state.doc.length, insert: value };
         }
```
Comment (Wroud) [inline webapp/packages/plugin-codemirror6/src/ReactCodemirror.tsx]:
```suggestion
        let isCursorInDoc = cursor && cursor.anchor > 0 && cursor.anchor < view.state.doc.length;

        if (value !== undefined && value !== view.state.doc.toString()) {
          transaction.changes = { from: 0, to: view.state.doc.length, insert: value };
          isCursorInDoc = cursor && cursor.anchor > 0 && cursor.anchor < value.length
        }
```

File: webapp/packages/plugin-sql-editor/src/SqlEditor/useSqlEditor.ts [hunk 3]
```diff
@@ -177,6 +177,16 @@ export function useSqlEditor(state: ISqlEditorTabState): ISQLEditorData {
       },
 
       setCursor(begin: number, end = begin): void {
+        if (begin > end) {
+          throw new Error('Cursor begin can not be greater than the end of it');
+        }
+
+        const script = this.parser.actualScript;
```
Comment (Wroud) [inline webapp/packages/plugin-sql-editor/src/SqlEditor/useSqlEditor.ts]:
```suggestion
        const scriptLength = this.value.length;
```

File: webapp/packages/plugin-sql-editor/src/SqlEditor/useSqlEditor.ts [hunk 4]
```diff
@@ -177,6 +177,16 @@ export function useSqlEditor(state: ISqlEditorTabState): ISQLEditorData {
...
+
+        if (end > script.length) {
+          end = script.length;
+        }
+
         this.cursor = {
           begin,
           end,
```
Comment (Wroud) [inline webapp/packages/plugin-sql-editor/src/SqlEditor/useSqlEditor.ts]:
```suggestion
        this.cursor = {
          begin: Math.min(begin, scriptLength),
          end: Math.min(end, scriptLength),
```

File: webapp/packages/plugin-sql-editor/src/SqlEditor/useSqlEditor.ts [hunk 5]
```diff
@@ -502,6 +512,10 @@ export function useSqlEditor(state: ISqlEditorTabState): ISQLEditorData {
     executor: data.dataSource?.onSetScript,
     handlers: [
       function setScript({ script }) {
+        if (data.cursor.end > script.length) {
+          data.setCursor(script.length);
+        }
```
Comment (Wroud) [inline webapp/packages/plugin-sql-editor/src/SqlEditor/useSqlEditor.ts]:
```suggestion
        // ensure that cursor is in script boundaries
        data.setCursor(data.cursor.begin, data.cursor.end);
```

File: webapp/packages/plugin-sql-editor/src/locales/ru.ts [hunk 6]
```diff
@@ -1,5 +1,7 @@
 export default [
   ['plugin_sql_editor_sql_editor_settings_group', 'SQL Редактор'],
+  ['sql_editor_settings_insert_table_aliases_name', 'Вставлять алиасы таблиц (в FROM условии)'],
+  ['sql_editor_settings_insert_table_aliases_desc', 'Automatically inserts table aliases'],
```
Comment (Wroud) [inline webapp/packages/plugin-sql-editor/src/locales/ru.ts]: please translate

**Review by Wroud — CHANGES_REQUESTED** [review]

File: webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts [hunk 7]
```diff
@@ -6,21 +6,34 @@
...
+import { SQL_EDITOR_SETTINGS_GROUP } from './SQL_EDITOR_SETTINGS_GROUP';
+
 const defaultSettings = schema.object({
   maxFileSize: schema.coerce.number().default(10 * 1024), // kilobyte
   disabled: schema.coerce.boolean().default(false),
   autoSave: schema.coerce.boolean().default(true),
 });
 
+const defaultProposalInsertTableSettings = schema.object({
+  alias: schema.coerce.boolean().default(true),
```
Comment (devnaumov) [inline webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts:28]: please use boolean name, `aliasesEnabled` or something like this

File: webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts [hunk 8]
```diff
@@ -6,21 +6,34 @@
...
+});
+
 export type SqlEditorSettings = schema.infer<typeof defaultSettings>;
 
 @injectable()
 export class SqlEditorSettingsService extends Dependency {
   readonly settings: PluginSettings<typeof defaultSettings>;
+  readonly proposalInsertTableSettings: PluginSettings<typeof defaultProposalInsertTableSettings>;
```
Comment (devnaumov) [inline webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts:36]: why do we need another settings object?
```
const defaultSettings = schema.object({
  maxFileSize: schema.coerce.number().default(10 * 1024), // kilobyte
  disabled: schema.coerce.boolean().default(false),
  autoSave: schema.coerce.boolean().default(true),
  aliasesEnabled: schema.coerce.boolean().default(true),
});
```

**Review by devnaumov — CHANGES_REQUESTED** [review]

File: webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts (hunk 7 above)
Comment (sergeyteleshev) [reply webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts:28]: `sql.proposals.insert.table.alias` - this is a pretty strict setting name I cannot rename cause it set on the backend side of the app. and it should be consistent to dbeaver desktop analogue setting

in the code of course it is used with the boolean prefix:
https://github.com/dbeaver/cloudbeaver/pull/2370/files#diff-603b6bb92ed26d86a6494be8e6ba52611dd550d3754f8bb9ed8b22ce7fa9ec81R42

File: webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts (hunk 8 above)
Comment (sergeyteleshev) [reply webapp/packages/plugin-sql-editor/src/SqlEditorSettingsService.ts:36]: cause `defaultSettings` has its own paths in product config
and new setting has another its own string path path which is not similar to `this.settings` at all

I tried to push the backend side to give us setting with the path like you want (so we have consistensy here) , but they said it is important to do it this way. And this is why we have this setting like this

**Review by Wroud — APPROVED** [review]

**Review by devnaumov — APPROVED** [review]

**Pull request merged**